Python/fileutils.c

```diff
@@ -1804,6 +1812,62 @@ _Py_write_impl(int fd, const void *buf, size_t count, int gil_held)
             }
         }
     }
+
+    // In non-blocking mode, a write that exceeds the available size of a pipe
+    // fails without writing anything. Limiting writes to the pipe size allows
+    // a buffered write to succeed eventually, as the pipe is read.
```

Comment:
Can we move this work until after we receive that error?

I'd prefer not to penalise users who stay below the buffer limit by making them do all this extra validation each time.

Reply:
At a minimum, can we refactor the main check out into its own function so that someone reading this bit can just see the name of what it's doing and not have to work through the logic twice.

Reply:
@zooba, if we retry the write, what do you think about iterating with half the count value until it's 1?

Reply:
It seems like we won't get any buffer size info back from the call, so that's really the only algorithm that makes sense. Would be great if we could raise a warning as well, but we don't really have any precedent for that. People can track on their own if they're regularly getting `n < len(data)` back from their write call.

Reply:
> It seems like we won't get any buffer size info back from the call, so that's really the only algorithm that makes sense. 

For a nonblocking pipe, a `WriteFile()` call that exceeds the available space in the pipe 'succeeds' with 0 bytes written. C `write()` handles this by failing with `ENOSPC`. AFAIK, the Windows API doesn't provide a way to query the available write quota. The internal query that's used by `GetNamedPipeInfo()` actually does return the available write quota, but the Windows API ignores it. 

Currently this PR caps the write at the pipe size, even if it's not all available. Eventually the reader will empty the pipe, and the full write will succeed. This works, albeit less aggressively than trying the write multiple times with a decreasing size until it succeeds. The latter is closer in overall behavior to Linux.

Reply:
Here's a simple example of the behavior when the write is retried in a loop with a decreasing count. Note that `CreatePipe()` uses a default pipe size of 4 KiB.

```pycon
>>> import os
>>> r, w = os.pipe()
>>> os.set_blocking(w, False)
>>> os.write(w, b'a' * 3072)
3072
>>> os.write(w, b'a' * 2048)
1024
>>> os.write(w, b'a' * 2048)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
BlockingIOError: [Errno 11] Resource temporarily unavailable
>>> os.read(r, 4096) == b'a' * 4096
True
```